File: pkg/bugnav/viewer/src/algorithms/mem_bfs.rs
//! Bounded-memory BFS — memory-aware pathfinding with a hard per-turn cost.
//!
//! Each turn:
//! 1. Sense (reveal cells within `r² ≤ VISION_R_SQ` from current pos).
//! 2. BFS from pos over the belief graph, capped at `BFS_BUDGET` cell
//!    expansions. Belief: in-bounds AND (undiscovered OR not-wall).
//!    Undiscovered is optimistic (treated as passable).
//! 3. Among cells reached by the bounded BFS, pick the one closest to
//!    goal. If strictly closer than current pos, step first cell along
//!    BFS path to it. Else fall back to Bug1.
//!
//! Worst-case per turn: O(BFS_BUDGET) — does not grow with map size or run
//! length. No caching, no replan bursts. This is the pattern production
//! bots use: fixed-budget local search, memory provides the passability
//! graph rather than amortizing work across turns.

use std::collections::{HashMap, HashSet, VecDeque};

use crate::algorithms::bug_common::{
    VISION_R_SQ, WallFollowState, WallStepOutcome, bresenham, dir_to_goal, dist_sq, has_los,
    neighbour, sensed_cells, wall_follow_step,
};
use crate::grid::Grid;
use crate::pathfinder::{Pathfinder, Snapshot, StepStatus};

/// Maximum cells BFS is allowed to expand per turn. Larger = better
/// plan quality; smaller = tighter WCET. 500 is ~20× the sensor window
/// and fits in ~sub-ms on a modern CPU.
const BFS_BUDGET: usize = 500;

const DIRS8: [(i32, i32); 8] = [
    (1, 0),
    (-1, 0),
    (0, 1),
    (0, -1),
    (1, 1),
    (1, -1),
    (-1, 1),
    (-1, -1),
];

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Mode {
    MotionToGoal,
    Circumnavigate,
    ReturnToLeave,
}

pub struct MemBfs {
    grid_w: i32,
    grid_h: i32,
    walls: Vec<bool>,
    goal: (i32, i32),
    pos: (i32, i32),
    discovered: Vec<bool>,
    // Bug1 state for fallback.
    mode: Mode,
    wf: WallFollowState,
    hit_point: (i32, i32),
    best_leave: (i32, i32),
    best_leave_dist_sq: i32,
    global_min_dist_sq: i32,
    follow_visited: HashSet<((i32, i32), (i32, i32), bool)>,
    steps: u32,
    snap: Snapshot,
    status: StepStatus,
}

pub fn build(grid: &Grid, start: (i32, i32), goal: (i32, i32)) -> Box<dyn Pathfinder> {
    let mut snap = Snapshot {
        current: start,
        path: vec![start],
        ..Snapshot::default()
    };
    snap.visited.insert(start);
    let discovered = vec![false; (grid.w * grid.h) as usize];
    Box::new(MemBfs {
        grid_w: grid.w,
        grid_h: grid.h,
        walls: grid.walls.clone(),
        goal,
        pos: start,
        discovered,
        mode: Mode::MotionToGoal,
        wf: WallFollowState {
            pos: start,
            current_obstacle: start,
            obstacle_on_right: true,
        },
        hit_point: start,
        best_leave: start,
        best_leave_dist_sq: dist_sq(start, goal),
        global_min_dist_sq: dist_sq(start, goal),
        follow_visited: HashSet::new(),
        steps: 0,
        snap,
        status: StepStatus::Running,
    })
}

impl MemBfs {
    fn idx(&self, x: i32, y: i32) -> Option<usize> {
        if x < 0 || y < 0 || x >= self.grid_w || y >= self.grid_h {
            None
        } else {
            Some((y * self.grid_w + x) as usize)
        }
    }

    fn sense(&mut self) {
        for c in sensed_cells(self.pos) {
            if let Some(i) = self.idx(c.0, c.1) {
                self.discovered[i] = true;
            }
        }
    }

    /// Belief passability: in-bounds AND (undiscovered OR not-wall).
    fn believed_passable(&self, x: i32, y: i32) -> bool {
        let Some(i) = self.idx(x, y) else { return false };
        !self.discovered[i] || !self.walls[i]
    }

    /// Ground-truth passability (only legitimate for cells within current
    /// sensor range — otherwise we're cheating). Used by Bug1 wall-follow,
    /// which only touches neighbour cells (always sensor-adjacent).
    fn passable(&self, x: i32, y: i32) -> bool {
        let Some(i) = self.idx(x, y) else { return false };
        !self.walls[i]
    }

    fn move_to(&mut self, new_pos: (i32, i32)) {
        self.pos = new_pos;
        self.snap.current = new_pos;
        self.snap.visited.insert(new_pos);
        self.snap.path.push(new_pos);
    }

    /// Bounded BFS over memory + optimistic-unknown, capped at BFS_BUDGET
    /// cell expansions. Returns the first-step cell of the path to the
    /// BFS-reachable cell closest to goal that is strictly closer than pos.
    /// Returns None if no such cell exists within the budget.
    fn try_bfs_step(&mut self, threshold: i32) -> bool {
        let next_step: Option<(i32, i32)> = {
            let mut queue: VecDeque<(i32, i32)> = VecDeque::new();
            let mut parent: HashMap<(i32, i32), (i32, i32)> = HashMap::new();
            queue.push_back(self.pos);
            parent.insert(self.pos, self.pos);
            let mut expansions = 0;
            let mut best: Option<((i32, i32), i32)> = None;
            while let Some(p) = queue.pop_front() {
                expansions += 1;
                if expansions > BFS_BUDGET {
                    break;
                }
                let d = dist_sq(p, self.goal);
                if d < threshold
                    && let Some((_, bd)) = best
                    && d < bd
                {
                    best = Some((p, d));
                } else if d < threshold && best.is_none() {
                    best = Some((p, d));
                }
                for (dx, dy) in DIRS8 {
                    let n = (p.0 + dx, p.1 + dy);
                    if !self.believed_passable(n.0, n.1) {
                        continue;
                    }
                    if parent.contains_key(&n) {
                        continue;
                    }
                    parent.insert(n, p);
                    queue.push_back(n);
                }
            }
            best.and_then(|(target, _)| {
                let mut cur = target;
                let mut next = target;
                while let Some(&p) = parent.get(&cur) {
                    if p == self.pos {
                        next = cur;
                        break;
                    }
                    if p == cur {
                        break; // reached pos (parent of itself)
                    }
                    cur = p;
                }
                if dist_sq(next, self.goal) < threshold {
                    Some(next)
                } else {
                    None
                }
            })
        };
        let Some(step) = next_step else {
            return false;
        };
        self.move_to(step);
        let d2 = dist_sq(self.pos, self.goal);
        if d2 < self.global_min_dist_sq {
            self.global_min_dist_sq = d2;
        }
        true
    }

    fn step_wall_follow(&mut self) -> WallStepOutcome {
        let (w, h, walls) = (self.grid_w, self.grid_h, &self.walls);
        let passable =
            |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h && !walls[(y * w + x) as usize];
        let on_map = |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h;
        let outcome = wall_follow_step(&mut self.wf, passable, on_map);
        if outcome == WallStepOutcome::Moved {
            self.move_to(self.wf.pos);
        }
        outcome
    }

    fn enter_circumnavigate(&mut self, blocked_dir: usize) {
        self.mode = Mode::Circumnavigate;
        self.wf = WallFollowState {
            pos: self.pos,
            current_obstacle: neighbour(self.pos, blocked_dir),
            obstacle_on_right: true,
        };
        self.hit_point = self.pos;
        self.best_leave = self.pos;
        self.best_leave_dist_sq = dist_sq(self.pos, self.goal);
        self.follow_visited.clear();
        self.follow_visited.insert((
            self.wf.pos,
            self.wf.current_obstacle,
            self.wf.obstacle_on_right,
        ));
    }
}

impl Pathfinder for MemBfs {
    fn step(&mut self) -> StepStatus {
        if self.status != StepStatus::Running {
            return self.status;
        }
        if self.pos == self.goal {
            self.status = StepStatus::Arrived;
            return self.status;
        }
        self.steps += 1;

        self.sense();

        // LoS shortcut — within sensor range, memory equals ground truth.
        let (w, h, walls) = (self.grid_w, self.grid_h, &self.walls);
        let passable_closure =
            move |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h && !walls[(y * w + x) as usize];
        if has_los(self.pos, self.goal, &passable_closure) {
            let line = bresenham(self.pos, self.goal);
            for &p in line.iter().skip(1) {
                self.move_to(p);
            }
            self.status = StepStatus::Arrived;
            return self.status;
        }

        let threshold = match self.mode {
            Mode::MotionToGoal => dist_sq(self.pos, self.goal),
            Mode::Circumnavigate | Mode::ReturnToLeave => self.best_leave_dist_sq,
        };
        if self.try_bfs_step(threshold) {
            if self.mode != Mode::MotionToGoal {
                self.mode = Mode::MotionToGoal;
                self.follow_visited.clear();
            }
            return self.status;
        }

        match self.mode {
            Mode::MotionToGoal => {
                let d = dir_to_goal(self.pos, self.goal);
                let np = neighbour(self.pos, d);
                if self.passable(np.0, np.1) {
                    self.move_to(np);
                    let d2 = dist_sq(self.pos, self.goal);
                    if d2 < self.global_min_dist_sq {
                        self.global_min_dist_sq = d2;
                    }
                } else {
                    self.enter_circumnavigate(d);
                }
            }
            Mode::Circumnavigate => {
                match self.step_wall_follow() {
                    WallStepOutcome::Moved => {}
                    WallStepOutcome::Surrounded => {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                }
                let d2 = dist_sq(self.pos, self.goal);
                if d2 < self.best_leave_dist_sq {
                    self.best_leave = self.pos;
                    self.best_leave_dist_sq = d2;
                }
                let state = (
                    self.wf.pos,
                    self.wf.current_obstacle,
                    self.wf.obstacle_on_right,
                );
                if !self.follow_visited.insert(state) {
                    if self.best_leave_dist_sq >= self.global_min_dist_sq {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                    self.global_min_dist_sq = self.best_leave_dist_sq;
                    self.mode = Mode::ReturnToLeave;
                    self.follow_visited.clear();
                    self.follow_visited.insert((
                        self.wf.pos,
                        self.wf.current_obstacle,
                        self.wf.obstacle_on_right,
                    ));
                }
            }
            Mode::ReturnToLeave => {
                if self.pos == self.best_leave {
                    self.mode = Mode::MotionToGoal;
                    return self.status;
                }
                match self.step_wall_follow() {
                    WallStepOutcome::Moved => {}
                    WallStepOutcome::Surrounded => {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                }
                let state = (
                    self.wf.pos,
                    self.wf.current_obstacle,
                    self.wf.obstacle_on_right,
                );
                if !self.follow_visited.insert(state) {
                    self.status = StepStatus::Unreachable;
                }
            }
        }
        self.status
    }

    fn snapshot(&self) -> &Snapshot {
        &self.snap
    }

    fn summary(&self) -> String {
        let discovered_count = self.discovered.iter().filter(|&&d| d).count();
        let total = self.discovered.len();
        format!(
            "pos: ({}, {})\ngoal: ({}, {})\nmode: {:?}\ndiscovered: {}/{} ({:.1}%)\nBFS budget: {}\nsensor r²≤{}\nsteps: {}\nstatus: {:?}",
            self.pos.0,
            self.pos.1,
            self.goal.0,
            self.goal.1,
            self.mode,
            discovered_count,
            total,
            100.0 * discovered_count as f64 / total as f64,
            BFS_BUDGET,
            VISION_R_SQ,
            self.steps,
            self.status,
        )
    }

    fn name(&self) -> &'static str {
        "Memory+BFS"
    }
}

File: pkg/bench_nav/src/bench_nav/stepped/bug/mem_bfs.py
"""Memory+BFS — stepped.

Maintains a per-query belief map (discovered cells + passable-neighbour
adjacency list) and runs a bounded BFS each step over the discovered cells.
Fallback to Bug1 circumnavigation when BFS fails to make progress.
"""

from __future__ import annotations

from enum import Enum
from typing import override

from bench_nav.common import INF
from bench_nav.precomputation import COST
from bench_nav.stepped.bug._common import (
    WallFollowState,
    WallStepOutcome,
    bresenham,
    dir_to_goal,
    dist_sq,
    has_los,
    make_passable_closures,
    neighbour,
    sensed_cells,
    wall_follow_step,
)
from bench_nav.types import PrecompCtx, Stepped

_BFS_BUDGET = 500
_DIST_INF = 65535
_DIRS8: tuple[tuple[int, int], ...] = (
    (1, 0),
    (-1, 0),
    (0, 1),
    (0, -1),
    (1, 1),
    (1, -1),
    (-1, 1),
    (-1, -1),
)


class _Mode(Enum):
    MOTION = 0
    CIRCUMNAV = 1
    RETURN_TO_LEAVE = 2


class MemBfs(Stepped):
    REQUIRES = frozenset({COST})

    @override
    def __init__(self, ctx: PrecompCtx) -> None:
        self.w = ctx.w
        self.h = ctx.h
        self.cost = ctx[COST]
        self.passable, self.on_map = make_passable_closures(
            ctx.w, ctx.h, self.cost, INF
        )
        self._active_goal: int | None = None
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(
            pos=(0, 0), current_obstacle=(0, 0), obstacle_on_right=True
        )
        self._best_leave: tuple[int, int] = (0, 0)
        self._best_leave_dist_sq = 0
        self._global_min_dist_sq = 0
        self._follow_visited: set[tuple[tuple[int, int], tuple[int, int], bool]] = set()
        self._los_queue: list[tuple[int, int]] = []
        n = ctx.w * ctx.h
        self._pnb: list[list[int]] = [[] for _ in range(n)]
        self._discovered = bytearray(n)
        self._bfs_dist: list[int] = [_DIST_INF] * n

    def _reset(self, p: tuple[int, int], g: tuple[int, int]) -> None:
        w, h = self.w, self.h
        n = w * h
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(pos=p, current_obstacle=p, obstacle_on_right=True)
        self._best_leave = p
        self._best_leave_dist_sq = dist_sq(p, g)
        self._global_min_dist_sq = dist_sq(p, g)
        self._follow_visited.clear()
        self._los_queue.clear()
        pnb = self._pnb
        for i in range(n):
            pnb[i] = []
        for y in range(h):
            for x in range(w):
                i = y * w + x
                row = pnb[i]
                for dx, dy in _DIRS8:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < w and 0 <= ny < h:
                        row.append(ny * w + nx)
        self._discovered = bytearray(n)

    def _sense(self, cur_pos: tuple[int, int]) -> None:
        w, h = self.w, self.h
        cost = self.cost
        discovered = self._discovered
        pnb = self._pnb
        for c in sensed_cells(cur_pos):
            x, y = c
            if not (0 <= x < w and 0 <= y < h):
                continue
            i = y * w + x
            if discovered[i]:
                continue
            discovered[i] = 1
            if cost[i] >= INF:
                for dx, dy in _DIRS8:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < w and 0 <= ny < h:
                        ni = ny * w + nx
                        if i in pnb[ni]:
                            pnb[ni].remove(i)
                pnb[i].clear()

    def _try_bfs_step(
        self, cur_pos: tuple[int, int], goal: tuple[int, int], threshold: int
    ) -> tuple[int, int] | None:
        w = self.w
        n = w * self.h
        pnb = self._pnb
        bfs_dist = self._bfs_dist
        pos_idx = cur_pos[1] * w + cur_pos[0]
        for i in range(n):
            bfs_dist[i] = _DIST_INF
        bfs_dist[pos_idx] = 0
        cur_frontier = [pos_idx]
        nxt_frontier: list[int] = []
        expansions = 0
        best_idx: int | None = None
        best_d = threshold
        reached_goal = False
        goal_idx = goal[1] * w + goal[0]

        while cur_frontier:
            for node in cur_frontier:
                expansions += 1
                if expansions > _BFS_BUDGET:
                    cur_frontier = []
                    nxt_frontier = []
                    break
                if node == goal_idx:
                    reached_goal = True
                    best_idx = node
                    cur_frontier = []
                    nxt_frontier = []
                    break
                nx, ny = node % w, node // w
                dx = nx - goal[0]
                dy = ny - goal[1]
                d = dx * dx + dy * dy
                if d < best_d:
                    best_d = d
                    best_idx = node
                next_d = bfs_dist[node] + 1
                for ni in pnb[node]:
                    if bfs_dist[ni] == _DIST_INF:
                        bfs_dist[ni] = next_d
                        nxt_frontier.append(ni)
            else:
                cur_frontier, nxt_frontier = nxt_frontier, []
                continue
            break

        if best_idx is None:
            return None
        cur = best_idx
        while bfs_dist[cur] > 1:
            cur_d = bfs_dist[cur]
            next_cell: int | None = None
            for ni in pnb[cur]:
                if bfs_dist[ni] == cur_d - 1:
                    next_cell = ni
                    break
            if next_cell is None:
                return None
            cur = next_cell
        if bfs_dist[cur] == _DIST_INF or cur == pos_idx:
            return None
        first_step = (cur % w, cur // w)
        if not reached_goal:
            fs_d = dist_sq(first_step, goal)
            if fs_d >= threshold:
                return None
        return first_step

    @override
    def step(self, pos: int, goal: int) -> int | None:
        w = self.w
        p = (pos % w, pos // w)
        g = (goal % w, goal // w)
        if goal != self._active_goal:
            self._active_goal = goal
            self._reset(p, g)
        if not self.passable(*p) or not self.passable(*g):
            return None
        if p == g:
            return pos

        if self._los_queue:
            nxt = self._los_queue.pop(0)
            return nxt[1] * w + nxt[0]
        self._sense(p)
        if has_los(p, g, self.passable):
            queue = bresenham(p, g)[1:]
            if queue:
                self._los_queue = queue[1:]
                first = queue[0]
                return first[1] * w + first[0]

        if self._mode is _Mode.MOTION:
            threshold = dist_sq(p, g)
        else:
            threshold = min(self._global_min_dist_sq, self._best_leave_dist_sq)
        bfs_step = self._try_bfs_step(p, g, threshold)
        if bfs_step is not None:
            d2 = dist_sq(bfs_step, g)
            self._global_min_dist_sq = min(self._global_min_dist_sq, d2)
            if self._mode is not _Mode.MOTION:
                self._mode = _Mode.MOTION
                self._follow_visited.clear()
            return bfs_step[1] * w + bfs_step[0]

        if self._mode is _Mode.MOTION:
            d = dir_to_goal(p, g)
            np = neighbour(p, d)
            if self.passable(*np):
                d2 = dist_sq(np, g)
                self._global_min_dist_sq = min(self._global_min_dist_sq, d2)
                return np[1] * w + np[0]
            self._mode = _Mode.CIRCUMNAV
            self._wf = WallFollowState(
                pos=p,
                current_obstacle=neighbour(p, d),
                obstacle_on_right=True,
            )
            self._best_leave = p
            self._best_leave_dist_sq = dist_sq(p, g)
            self._follow_visited.clear()
            self._follow_visited.add(
                (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
            )
            return pos
        if self._mode is _Mode.CIRCUMNAV:
            self._wf.pos = p
            outcome = wall_follow_step(self._wf, self.passable, self.on_map)
            if outcome is WallStepOutcome.SURROUNDED:
                return None
            np = self._wf.pos
            d2 = dist_sq(np, g)
            if d2 < self._best_leave_dist_sq:
                self._best_leave = np
                self._best_leave_dist_sq = d2
            state = (
                self._wf.pos,
                self._wf.current_obstacle,
                self._wf.obstacle_on_right,
            )
            if state in self._follow_visited:
                if self._best_leave_dist_sq >= self._global_min_dist_sq:
                    return None
                self._global_min_dist_sq = self._best_leave_dist_sq
                self._mode = _Mode.RETURN_TO_LEAVE
                self._follow_visited.clear()
                self._follow_visited.add(state)
            else:
                self._follow_visited.add(state)
            return np[1] * w + np[0]
        # RETURN_TO_LEAVE
        if p == self._best_leave:
            self._mode = _Mode.MOTION
            return pos
        self._wf.pos = p
        outcome = wall_follow_step(self._wf, self.passable, self.on_map)
        if outcome is WallStepOutcome.SURROUNDED:
            return None
        np = self._wf.pos
        state = (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
        if state in self._follow_visited:
            return None
        self._follow_visited.add(state)
        return np[1] * w + np[0]
